src/lib.rs

```diff
@@ -1,602 +1,62 @@
 #![deny(rust_2018_idioms)]
 
-use ordermap::{orderset, OrderMap, OrderSet};
-use std::collections::hash_map::Entry;
+use crate::context::{Context, IStr};
+use indexmap::IndexMap;
 use std::collections::HashMap;
 use std::hash::Hash;
-use std::iter;
-use std::ops::{Add, BitAnd, BitOr};
-use std::rc::Rc;
 
-pub struct Grammar<Pat> {
-    pub rules: OrderMap<String, RuleWithNamedFields<Pat>>,
+pub mod context;
+pub mod rule;
+
+pub struct Grammar {
```

Comment:
Putting a note here since this threw me off initially:

The grammar does not own the context such that the context can easily be reused between multiple `Grammar`s. This is important when often recreating the `Grammar` (such as in an interactive session), or when building a `Grammar` from multiple smaller `Grammar`s (such as our many-file system in wg-grammar).

Reply:
I've been meaning to rename this type to `Module` or something, because that's more or less what it is. Not sure if I should do it now or later though.

Reply:
Leave it for now, we can always rename later.